src/gcp/secretManager.ts

```diff
@@ -63,123 +94,131 @@ export async function secretExists(projectId: string, name: string): Promise<boo
 }
 
 export async function createSecret(
   projectId: string,
   name: string,
   labels: Record<string, string>
 ): Promise<Secret> {
-  const createRes = await api.request(
-    "POST",
-    `/v1beta1/projects/${projectId}/secrets?secretId=${name}`,
+  const createRes = await client.post<CreateSecretRequest, Secret>(
+    `projects/${projectId}/secrets?secretId=${name}`,
     {
-      auth: true,
-      origin: api.secretManagerOrigin,
-      data: {
-        replication: {
-          automatic: {},
-        },
-        labels,
+      replication: {
+        automatic: {},
       },
+      labels,
     }
   );
   return parseSecretResourceName(createRes.body.name);
 }
 
-export async function addVersion(secret: Secret, payloadData: string): Promise<SecretVersion> {
-  const res = await api.request(
-    "POST",
-    `/v1beta1/projects/${secret.projectId}/secrets/${secret.name}:addVersion`,
+export async function addVersion(
+  secret: Secret,
+  payloadData: string
+): Promise<Required<SecretVersion>> {
+  const res = await client.post<AddVersionRequest, { name: string; state: SecretVersionState }>(
+    `projects/${secret.projectId}/secrets/${secret.name}:addVersion`,
     {
-      auth: true,
-      origin: api.secretManagerOrigin,
-      data: {
-        payload: {
-          data: Buffer.from(payloadData).toString("base64"),
-        },
+      payload: {
+        data: Buffer.from(payloadData).toString("base64"),
       },
     }
   );
-  const nameTokens = res.body.name.split("/");
   return {
-    secret: {
-      projectId: nameTokens[1],
-      name: nameTokens[3],
-    },
-    versionId: nameTokens[5],
+    ...parseSecretVersionResourceName(res.body.name),
+    state: res.body.state,
   };
 }
 
-export async function grantServiceAgentRole(
+export async function getIamPolicy(secret: Secret): Promise<iam.Policy> {
+  const res = await client.get<iam.Policy>(
+    `projects/${secret.projectId}/secrets/${secret.name}:getIamPolicy`
+  );
+  return res.body;
+}
+
+export async function setIamPolicyBindings(secret: Secret, bindings: iam.Binding[]): Promise<void> {
+  await client.post<{ policy: Partial<iam.Policy> }, iam.Policy>(
+    `projects/${secret.projectId}/secrets/${secret.name}:setIamPolicy`,
+    {
+      policy: {
+        bindings,
+      },
+    },
+    {
+      queryParams: {
+        updateMask: "bindings",
+      },
+    }
+  );
+}
+
+export async function ensureServiceAgentRole(
   secret: Secret,
-  serviceAccountEmail: string,
+  serviceAccountEmails: string[],
   role: string
 ): Promise<void> {
-  const getPolicyRes = await api.request(
-    "GET",
-    `/v1beta1/projects/${secret.projectId}/secrets/${secret.name}:getIamPolicy`,
-    {
-      auth: true,
-      origin: api.secretManagerOrigin,
+  const policy = await module.exports.getIamPolicy(secret);
+  const bindings = policy.bindings || [];
+
+  const newBindings = [];
+  for (const serviceAccountEmail of serviceAccountEmails) {
```

Comment:
Rather than adding one binding per service account, find or create a binding for the required role and add all service accounts as members. You can see how Cole did this in the functions APIs for the invoker role.

Reply:
I can see why we might want to do it (it's strictly more correct?), but IAM endpoint seems happily collate the bindings into one for me. Do you prefer we don't rely on this behavior?

Reply:
This is now fixed.